File: packages/core/api/endpoints/chat-links.ts
import { z } from "zod";
import { TaskSchema, type Task } from "../../types/task";
import { request } from "../http";
import { parseWithFallback } from "../schema";
import { enc } from "./chat-schemas";

const ChatMessageLinkSchema = z.object({
  id: z.string(),
  message_id: z.string(),
  target_type: z.string(),
  target_id: z.string(),
  relation: z.string().optional().default(""),
  created_by: z.string().optional().default(""),
  created_at: z.string().optional().default(""),
});

const LinkEnvelopeSchema = z.object({
  link: ChatMessageLinkSchema.optional(),
});

const LinkListSchema = z.object({
  links: z.array(ChatMessageLinkSchema).optional().default([]),
});

const TaskEnvelopeSchema = z.object({
  task: TaskSchema.optional(),
});

const OkSchema = z.object({
  ok: z.boolean().optional(),
  status: z.string().optional(),
});

export type ChatMessageLinkRecord = z.infer<typeof ChatMessageLinkSchema>;

export type CreateTaskFromMessageInput = {
  title?: string;
  project_id?: string;
  assignee_id?: string;
  assignee_kind?: string;
  due_date?: string;
  sync_thread?: boolean;
  priority?: string;
};

export type CreateChatMessageLinkInput = {
  target_type: string;
  target_id: string;
};

export type SyncThreadTaskInput = {
  task_id: string;
  direction?: string;
};

function isOk(parsed: { ok?: boolean; status?: string }): boolean {
  return parsed.ok === true || parsed.status === "ok";
}

export async function createTaskFromChatMessage(
  workspaceId: string,
  messageId: string,
  input: CreateTaskFromMessageInput = {},
): Promise<Task | null> {
  const raw = await request(
    `/api/v1/workspaces/${enc(workspaceId)}/chat/messages/${enc(messageId)}/tasks`,
    { method: "POST", body: input },
  );
  const parsed = parseWithFallback(raw, TaskEnvelopeSchema, { task: undefined }, {
    endpoint: "POST .../chat/messages/{id}/tasks",
  });
  return parsed.task ?? null;
}

export async function listChatMessageLinks(
  workspaceId: string,
  messageId: string,
): Promise<ChatMessageLinkRecord[]> {
  const raw = await request(
    `/api/v1/workspaces/${enc(workspaceId)}/chat/messages/${enc(messageId)}/links`,
  );
  const parsed = parseWithFallback(raw, LinkListSchema, { links: [] }, {
    endpoint: "GET .../chat/messages/{id}/links",
  });
  return parsed.links ?? [];
}

export async function createChatMessageLink(
  workspaceId: string,
  messageId: string,
  input: CreateChatMessageLinkInput,
): Promise<ChatMessageLinkRecord | null> {
  const raw = await request(
    `/api/v1/workspaces/${enc(workspaceId)}/chat/messages/${enc(messageId)}/links`,
    { method: "POST", body: input },
  );
  const parsed = parseWithFallback(raw, LinkEnvelopeSchema, { link: undefined }, {
    endpoint: "POST .../chat/messages/{id}/links",
  });
  return parsed.link ?? null;
}

export async function deleteChatMessageLink(
  workspaceId: string,
  messageId: string,
  linkId: string,
): Promise<boolean> {
  const raw = await request(
    `/api/v1/workspaces/${enc(workspaceId)}/chat/messages/${enc(messageId)}/links/${enc(linkId)}`,
    { method: "DELETE" },
  );
  const parsed = parseWithFallback(raw, OkSchema, {}, {
    endpoint: "DELETE .../chat/messages/{id}/links/{linkID}",
  });
  return isOk(parsed);
}

export async function syncChatThreadTask(
  workspaceId: string,
  threadRootId: string,
  input: SyncThreadTaskInput,
): Promise<boolean> {
  const raw = await request(
    `/api/v1/workspaces/${enc(workspaceId)}/chat/threads/${enc(threadRootId)}/task-sync`,
    { method: "POST", body: input },
  );
  const parsed = parseWithFallback(raw, OkSchema, {}, {
    endpoint: "POST .../chat/threads/{id}/task-sync",
  });
  return isOk(parsed);
}

export async function unsyncChatThreadTask(
  workspaceId: string,
  threadRootId: string,
): Promise<boolean> {
  const raw = await request(
    `/api/v1/workspaces/${enc(workspaceId)}/chat/threads/${enc(threadRootId)}/task-sync`,
    { method: "DELETE" },
  );
  const parsed = parseWithFallback(raw, OkSchema, {}, {
    endpoint: "DELETE .../chat/threads/{id}/task-sync",
  });
  return isOk(parsed);
}

File: server/internal/handler/chat_links.go
package handler

import (
	"net/http"
	"time"

	"github.com/go-chi/chi/v5"

	"github.com/unicomhub/uniwork/server/internal/handler/dto/sdi"
	"github.com/unicomhub/uniwork/server/internal/handler/dto/sdo"
	"github.com/unicomhub/uniwork/server/internal/middleware"
	"github.com/unicomhub/uniwork/server/internal/service"
)

func (h *handlers) createTaskFromChatMessage(w http.ResponseWriter, r *http.Request) {
	var in sdi.CreateTaskFromMessageSDI
	if !decode(w, r, &in, maxJSONBody) {
		return
	}
	task, err := h.Chat.CreateTaskFromMessage(
		r.Context(),
		middleware.UserID(r.Context()),
		chi.URLParam(r, "workspaceID"),
		chi.URLParam(r, "messageID"),
		service.CreateTaskFromMessageInput{
			Title: in.Title, ProjectID: in.ProjectID, AssigneeID: in.AssigneeID,
			AssigneeKind: in.AssigneeKind, DueDate: in.DueDate, SyncThread: in.SyncThread,
			Priority: in.Priority,
		},
	)
	if err != nil {
		h.mapServiceError(w, err)
		return
	}
	h.respondTask(w, r, http.StatusCreated, task)
}

func (h *handlers) createChatMessageLink(w http.ResponseWriter, r *http.Request) {
	var in sdi.CreateChatMessageLinkSDI
	if !decode(w, r, &in, maxJSONBody) {
		return
	}
	link, err := h.Chat.LinkChatMessage(
		r.Context(),
		middleware.UserID(r.Context()),
		chi.URLParam(r, "workspaceID"),
		chi.URLParam(r, "messageID"),
		service.CreateChatMessageLinkInput{TargetType: in.TargetType, TargetID: in.TargetID},
	)
	if err != nil {
		h.mapServiceError(w, err)
		return
	}
	respondJSON(w, http.StatusOK, map[string]any{"link": toChatMessageLinkDTO(link)})
}

func (h *handlers) listChatMessageLinks(w http.ResponseWriter, r *http.Request) {
	rows, err := h.Chat.ListChatMessageLinks(
		r.Context(),
		middleware.UserID(r.Context()),
		chi.URLParam(r, "workspaceID"),
		chi.URLParam(r, "messageID"),
	)
	if err != nil {
		h.mapServiceError(w, err)
		return
	}
	out := make([]sdo.ChatMessageLinkDTO, 0, len(rows))
	for _, row := range rows {
		out = append(out, toChatMessageLinkDTO(row))
	}
	respondJSON(w, http.StatusOK, sdo.ChatMessageLinkListSDO{Links: out})
}

func (h *handlers) deleteChatMessageLink(w http.ResponseWriter, r *http.Request) {
	if err := h.Chat.UnlinkChatMessage(
		r.Context(),
		middleware.UserID(r.Context()),
		chi.URLParam(r, "workspaceID"),
		chi.URLParam(r, "messageID"),
		chi.URLParam(r, "linkID"),
	); err != nil {
		h.mapServiceError(w, err)
		return
	}
	respondJSON(w, http.StatusOK, map[string]any{"ok": true})
}

func (h *handlers) syncChatThreadTask(w http.ResponseWriter, r *http.Request) {
	var in sdi.SyncThreadTaskSDI
	if !decode(w, r, &in, maxJSONBody) {
		return
	}
	if _, err := h.Chat.SyncThreadTask(
		r.Context(),
		middleware.UserID(r.Context()),
		chi.URLParam(r, "workspaceID"),
		chi.URLParam(r, "messageID"),
		service.SyncThreadTaskInput{TaskID: in.TaskID, Direction: in.Direction},
	); err != nil {
		h.mapServiceError(w, err)
		return
	}
	respondJSON(w, http.StatusOK, map[string]any{"ok": true})
}

func (h *handlers) unsyncChatThreadTask(w http.ResponseWriter, r *http.Request) {
	if err := h.Chat.UnsyncThreadTask(
		r.Context(),
		middleware.UserID(r.Context()),
		chi.URLParam(r, "workspaceID"),
		chi.URLParam(r, "messageID"),
	); err != nil {
		h.mapServiceError(w, err)
		return
	}
	respondJSON(w, http.StatusOK, map[string]any{"ok": true})
}

func toChatMessageLinkDTO(row service.ChatMessageLinkRow) sdo.ChatMessageLinkDTO {
	return sdo.ChatMessageLinkDTO{
		ID: row.ID, MessageID: row.MessageID, TargetType: row.TargetType,
		TargetID: row.TargetID, Relation: row.Relation, CreatedBy: row.CreatedBy,
		CreatedAt: row.CreatedAt.UTC().Format(time.RFC3339),
	}
}
